cassandra-scheduler/src/main/java/io/mesosphere/mesos/frameworks/cassandra/CassandraCluster.java

```diff
@@ -206,99 +285,200 @@ public void recordHealthCheck(@NotNull final String executorId, @NotNull final C
+
+            if (!result.hasAnyTask())
+                // nothing to do
+                return null;
+
             final CassandraNode built = node.build();
             clusterState.addOrSetNode(built);
-            final Tuple2<CassandraNodeExecutor, List<CassandraNodeTask>> retVal = tuple2(built.getCassandraNodeExecutor(), tasks);
-            return Optional.of(retVal);
-        } else if (launchNode()) {
-            clusterState.nodes(append(
-                    clusterState.nodes(),
-                    buildCassandraNode(offer)
-            ));
-            return Optional.absent();
-        } else {
-            return Optional.absent();
+
+            return result;
+        } finally {
+            if (LOGGER.isDebugEnabled())
+                LOGGER.debug(marker, "< getTasksForOffer(offer : {}) = {}, {}", protoToString(offer));
         }
     }
 
-    private static CassandraNode buildCassandraNode(Protos.Offer offer) {
+    private boolean canLaunchServerTask() {
+        return clock.now().getMillis() > nextPossibleServerLaunchTimestamp();
+    }
+
+    public long nextPossibleServerLaunchTimestamp() {
+        long lastServerLaunchTimestamp = getClusterState().get().getLastServerLaunchTimestamp();
+        long seconds = Math.max(getConfiguration().get().getBootstrapGraceTimeSeconds(), getConfiguration().get().getHealthCheckIntervalSeconds());
+        return lastServerLaunchTimestamp + seconds * 1000L;
+    }
+
+    @NotNull
+    public Optional<CassandraNode> cassandraNodeForHostname(String hostname) {
+        return headOption(
+            from(clusterState.nodes())
+                .filter(cassandraNodeHostnameEq(hostname))
+        );
+    }
+
+    @NotNull
+    public Optional<CassandraNode> cassandraNodeForExecutorId(String executorId) {
+        return headOption(
+                from(clusterState.nodes())
+                        .filter(cassandraNodeExecutorIdEq(executorId))
+        );
+    }
+
+    private CassandraNode buildCassandraNode(Protos.Offer offer, boolean seed) {
         CassandraNode.Builder builder = CassandraNode.newBuilder()
-                .setHostname(offer.getHostname());
+                .setHostname(offer.getHostname())
+                .setSeed(seed);
         try {
-            int jmxPort = defaultCassandraPortMappings.get("jmx_port").intValue();
             InetAddress ia = InetAddress.getByName(offer.getHostname());
-//            if (ia.isLoopbackAddress())
-//                try (ServerSocket serverSocket = new ServerSocket(0)) {
-//                    jmxPort = serverSocket.getLocalPort();
-//                } catch (IOException e) {
-//                    throw new RuntimeException(e);
-//                }
+
+            int jmxPort = getPortMapping(PORT_JMX);
+            if (ia.isLoopbackAddress()) {
+                try (ServerSocket serverSocket = new ServerSocket(0)) {
```

Comment:
Instead of picking a port from the host, a port from the Offer should be used (especially because this port is being evaluated on the node running the framework and not necessarily the node running the server)